File: 实现代码/expert/research-retro-assistant/bin/rra/library/store.py
"""库的读写与导入合并。编号冲突时重编号并记录 remap_from，不静默覆盖。"""

import json
from dataclasses import dataclass, field
from pathlib import Path

from ..contracts.models import Archive, Library
from ..contracts.validator import Validator
from .dedup import Deduplicator


@dataclass
class MergeReport:
    """导入结果：新增、重复、重编号各自的条数与明细。"""

    added: list[str] = field(default_factory=list)
    duplicated: list[str] = field(default_factory=list)
    renumbered: dict[str, field_factory := dict] = field(default_factory=dict) if False else field(default_factory=dict)
    violations: list[str] = field(default_factory=list)


class LibraryStore:
    """经验库的存取。路径由调用方给出，不硬编码。"""

    def __init__(self, path: Path) -> None:
        self.path = Path(path)
        self.dedup = Deduplicator()
        self.validator = Validator()

    def load(self) -> Library:
        raw = json.loads(self.path.read_text(encoding="utf-8"))
        vs = self.validator.validate_library(raw)
        if vs:
            raise ValueError("库文件未通过契约校验：" + "；".join(str(v) for v in vs))
        return Library.from_dict(raw)

    def save(self, library: Library) -> None:
        raw = library.to_dict()
        vs = self.validator.validate_library(raw)
        if vs:
            raise ValueError("拒绝写入未通过校验的库：" + "；".join(str(v) for v in vs))
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.path.write_text(
            json.dumps(raw, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    def next_id(self, library: Library) -> str:
        """按 R-### 规则分配下一个编号：现有最大序号 + 1。"""
        best = 0
        for r in library.records:
            try:
                best = max(best, int(r.id.split("-")[1]))
            except (IndexError, ValueError):
                continue
        return "R-%03d" % (best + 1)

    def merge(self, base: Library, incoming: Library) -> tuple[Library, MergeReport]:
        """合并：按指纹去重；编号冲突且指纹不同则重编号并记录 remap_from。"""
        report = MergeReport()
        merged = Library(version=base.version, records=list(base.records),
                         generated_at=base.generated_at)
        known_keys = {self.dedup.key_of(r) for r in merged.records}
        known_ids = {r.id for r in merged.records}

        for rec in incoming.records:
            key = self.dedup.key_of(rec)
            if key in known_keys:
                report.duplicated.append(rec.id)
                continue
            known_keys.add(key)
            if rec.id in known_ids:
                new_id = self.next_id(merged)
                report.renumbered[rec.id] = new_id
                data = rec.to_dict()
                data["id"] = new_id
                data["provenance"]["remap_from"] = rec.id
                rec = Archive.from_dict(data)
            known_ids.add(rec.id)
            merged.records.append(rec)
            report.added.append(rec.id)

        return merged, report

    def find_abandoned_with_blocker(self) -> list[Archive]:
        """列出「已放弃且阻塞点明确」的档案——死实验复活的输入。

        阻塞点明确 = blocker 非空且有 missing_info 之外的明确描述；
        这里以非空为最低门槛，具体筛法由 ResurrectionSkill 补充。
        """
        lib = self.load()
        return [r for r in lib.records if r.status == "已放弃" and r.blocker.strip()]

File: 实现代码/web/js/store.js
import { PersistenceAdapter } from "./persist.js";
import { ContractValidator } from "./contracts.js";
import { SAMPLE_LIBRARY } from "../data/library.sample.js";

const SEP = "\u001f";

/** 「人工纠错」标记与目标编号正则：必须与 src/rra/library/refutation.py、src/rra/library/dedup.py 逐字一致。 */
const REF_MARKER = "人工纠错";
const REF_ID = /R-\d+/;

function isRefutation(rec) {
  return !!(rec && typeof rec.attempt === "string" && rec.attempt.includes(REF_MARKER));
}

/** 纠错记录专用键 = 标记 + 目标编号 + 归一化理由；取不到目标编号时返回 null（退回通用键）。 */
function refutationKey(rec, norm) {
  if (!isRefutation(rec)) return null;
  let target = "";
  for (const lk of (rec.links || [])) {
    if (lk && lk.relation === "冲突") { target = lk.target || ""; break; }
  }
  if (!target) { const m = REF_ID.exec(rec.attempt || ""); if (m) target = m[0]; }
  if (!target) return null;
  const reason = rec.observation || ((rec.attribution && rec.attribution.text) || "");
  return [norm(REF_MARKER), norm(target), norm(reason)].join(SEP);
}

/** 与 Python 侧 Deduplicator.key_of 逐字对齐：尝试 / 阻塞点 / 关键条件 / 关联产物，后两段排序。
 *  纠错记录走专用键（同目标不同理由不再撞键），普通记录语义一个字不改。 */
function dedupKey(rec) {
  const norm = (s) => String(s === undefined || s === null ? "" : s)
    .normalize("NFKC")
    .replace(/[，。！？、；：,.!?;:"'“”‘’（）()\[\]【】]+/g, "")
    .replace(/\s+/g, "")
    .toLowerCase();
  const ref = refutationKey(rec, norm);
  if (ref !== null) return ref;
  const conditions = [];
  for (const lk of (rec.links || [])) {
    for (const s of (lk.same || [])) conditions.push(s.dim + "=" + norm(s.value));
    for (const d of (lk.diff || [])) conditions.push(d.dim + ":" + norm(d.from) + "->" + norm(d.to));
  }
  const artifacts = (rec.artifacts || [])
    .filter(a => a && typeof a === "object")
    .map(a => norm(a.ref));
  const parts = [norm(rec.attempt), norm(rec.blocker),
    conditions.slice().sort().join(","),
    artifacts.slice().sort().join(",")].join(SEP);
  const own = Object.entries(rec.conditions || {}).map(([k,v])=>[k,String(v).normalize('NFKC').trim().replace(/\s+/g,' ').toLowerCase()]).sort(([a],[b])=>a<b?-1:a>b?1:0);
  return parts + (own.length ? SEP + JSON.stringify(own) : '');
}

/** 库状态与导入导出。合并规则与 Python 侧一致（同一契约、同一套去重键）。 */
export class LibraryStore {
  constructor(adapter = new PersistenceAdapter(), validator = new ContractValidator()) {
    this.adapter = adapter;
    this.validator = validator;
    this.library = { version: "0.1.0", generated_at: "", records: [] };
    this.listeners = [];
  }

  async load() {
    const local = this.adapter.read("library");
    if (local) {
      const issues=this.validator.validateLibrary(local);
      if(issues.length) throw new Error('本地库校验失败；请先导出备份并检查档案。');
      // Only upgrade untouched demonstration records; never infer conditions for user records.
      let changed=false;
      for(const r of local.records){
        const seed=SAMPLE_LIBRARY.records.find(s=>s.id===r.id);
        if(!r.conditions&&seed?.conditions&&r.provenance?.source==='模拟'){
          const withoutConditions=s=>JSON.stringify(Object.fromEntries(Object.entries(s).filter(([k])=>k!=='conditions').sort(([a],[b])=>a.localeCompare(b))));
          if(withoutConditions(r)===withoutConditions(seed)){r.conditions={...seed.conditions};changed=true;}
        }
      }
      if(changed)this.adapter.write('library',local);
      this.library = local;
      return this.library;
    }
    try {
      const res = await fetch("./data/library.sample.json");
      if (!res.ok) throw new Error('样例读取失败');
      this.library = await res.json();
    } catch {
      // fetch 不可用（file:// 打开 / 服务器未起）时回退到内联样例，
      // 保证「双击 index.html 也能看到库」，而不是白屏。
      this.library = SAMPLE_LIBRARY;
    }
    return this.library;
  }

  /**
   * 提交一个新库：**先落盘成功，才替换内存**。
   *
   * 为什么要这样（2026-09-19 外部审查 P1）：原实现在 importLibrary 里先 `this.library = next`
   * 再 `save()`。一旦 save 因配额/隐私模式抛错，内存已经是新库、存储还是旧库 ——
   * 调用方以为"失败了什么都没变"，实际内存被污染，之后任何一次保存都会把半截状态写出去。
   */
  commit(next) {
    const vs = this.validator.validateLibrary(next);
    if (vs.length) throw new Error("库未通过契约校验：" + vs.map(v => v.path).join("；"));
    this.adapter.write("library", next);      // 写失败会抛错，此时内存保持旧值
    this.library = next;
    this.lastWrite = JSON.stringify(next);
    this.emit();
    return next;
  }

  /** 保存当前内存态（等价于提交自己）。 */
  save() {
    return this.commit(this.library);
  }

  /** 其它标签页改动了库：整库替换后通知视图（跳过自己刚写的那次，避免回环）。 */
  adoptExternal(raw) {
    if (!raw || !Array.isArray(raw.records)) return false;
    const text = JSON.stringify(raw);
    if (text === this.lastWrite) return false;
    this.library = raw;
    this.emit();
    return true;
  }

  onChange(fn) { this.listeners.push(fn); }
  emit() { this.listeners.forEach(fn => fn(this.library)); }

  /** 导入：合并、去重、冲突重编号并记录 remap_from，返回前后对比提示。 */
  importLibrary(rawJson) {
    let incoming;
    try { incoming = typeof rawJson === "string" ? JSON.parse(rawJson) : rawJson; }
    catch (e) { throw new Error("不是合法 JSON：" + e.message); }
    const vs = this.validator.validateLibrary(incoming);
    if (vs.length) throw new Error("导入文件未通过契约校验：" + vs.map(v => v.path).join("；"));

    // 先算出合并结果，全部成功后再整体替换。
    // 原实现边 push 边 save：save 校验失败时内存已被改脏，而调用方以为"内存未受影响"。
    const report = { added: [], duplicated: [], renumbered: {} };
    const next = JSON.parse(JSON.stringify(this.library));
    const knownKeys = new Set(next.records.map(dedupKey));
    const knownIds = new Set(next.records.map(r => r.id));
    const maxNo = () => Math.max(0, ...[...knownIds].map(id => parseInt(id.slice(2), 10) || 0));

    for (const rec of incoming.records) {
      const key = dedupKey(rec);
      if (knownKeys.has(key)) { report.duplicated.push(rec.id); continue; }
      knownKeys.add(key);
      const clone = JSON.parse(JSON.stringify(rec));
      if (knownIds.has(clone.id)) {
        const newId = "R-" + String(maxNo() + 1).padStart(3, "0");
        report.renumbered[clone.id] = newId;
        clone.id = newId;
        clone.provenance = clone.provenance || {};
        clone.provenance.remap_from = rec.id;
      }
      knownIds.add(clone.id);
      next.records.push(clone);
      report.added.push(clone.id);
    }

    const postVs = this.validator.validateLibrary(next);
    if (postVs.length) {
      throw new Error("导入后库未通过契约校验（已回滚，内存未改动）："
        + postVs.map(v => v.path).join("；"));
    }
    this.commit(next);        // 先落盘、成功才替换内存；写失败时内存保持旧库
    return report;
  }

  exportLibrary() {
    return JSON.stringify(this.library, null, 2);
  }

  findById(id) {
    return this.library.records.find(r => r.id === id) || null;
  }

  /** 被关联次数：派生指标，从 links[].target 统计，不写入档案。 */
  linkedCount(id) {
    return this.library.records.reduce(
      (n, r) => n + (r.links || []).filter(l => l.target === id).length, 0);
  }
}
